haskell-api.md

```diff
@@ -260,40 +260,44 @@ Example:
         / StudentCoursesR GET   --> Handlers/V3/Students/Student/Courses.hs
 ```
 
-## Module Structure
+## API Types
 
-- Prefer in-module request/response types
+Prefer distinct types for requests and responses that are only used for that. We
+call these "API Types". Do not use persistent Entity types as API request or
+response types.
 
-  - Request type naming: `{Resource}{Method}Body`
-  - Response type naming: `{Resource}{Method}`
-  - NOTE: list responses are `[{Resource}{Method}]`, not `{Resources}{Method}`
+1. A response API Type should be named `Api{Resource}` and should be re-used
+   anywhere `Resource` appears in the API.
 
-  ```hs
-  -- Good
-  data TeacherGet
+   If the same concept takes on a different shape in different places, it should
+   be considered a different Resource, with a different name. For example,
+   `ApiTeacher`, `ApiAbbreviatedTeacher`, `ApiTeacherWithSchool`,
+   `ApiUsageReportTeacher`.
 
-  getTeachersR :: Handler [TeacherGet]
-  getTeachersR = undefined
+1. A request API Type should be named `Api{Action}{Resource}({Target})`.
 
-  data TeacherPost
-  data TeacherPostBody
+   ```hs
+   -- Good (Target omitted, acting on entire resource)
+   data ApiCreateTeacher
+   data ApiDeleteTeacher
 
-  postTeacherR :: Handler TeacherPost
-  postTeacherR = do
-    body <- requireJsonBody @TeacherPostBody
+   -- Good (Target included)
+   data ApiSetTeacherPassword
 
-  -- Bad
-  data Teachers
+   -- Bad
+   data TeacherPUT
+   data CreateApiTeacher
+   data ApiTeacherSetPassword
+   ```
 
-  getTeachersR :: Handler Teachers
-  getTeachersR = undefined
+1. An API Type should not be exported unless in a module of only it (because
```

Comment:
Cosmetic / nitpick. What do you think of: 

1. An API Type should not be exported unless it is shared and in its own module.

Reply:
Are you suggesting dropping the entire parenthetical?

Reply:
Nope; I am finding the current wording a bit "off", particularly the "unless in a module of only it".

Reply:
How is it now?